kubectl debug: Add --keep-* flag to control the removal of probes, labels, annotations, and initContainers from copy pod

mochizuki875 opened the pull request:
#### What type of PR is this?

/kind feature
/cleanup

#### What this PR does / why we need it:

When using `--copy-to` option with some debugging profile(legacy, restricted, netadmin sysadmin), the probe configuration remains in the copy pod .
Therefore, if the probe condition is not met by debugging operation on the copy pod, the main container will be restarted.
This may prevent debugging.

So in this PR, I've added flags(`--keep-liveness`,`--keep-readiness`,`--keep-startup`) to control the removal of probes from copy pod.
In addition, I've added `--keep-labels`, `--keep-annotations` and `--keep-init-containers` to control the removal of labels, annotations and initContainers too.


#### Which issue(s) this PR fixes:

#### Special notes for your reviewer:

#### Does this PR introduce a user-facing change?

```release-note
Add `--keep-*` flags to `kubectl debug`, which enables to control the removal of probes, labels, annotations and initContainers from copy pod.
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

```docs
```

mochizuki875 committed:
Add keep options

mochizuki875 commented:
/sig cli

ardaguclu commented on staging/src/k8s.io/kubectl/pkg/cmd/debug/profiles.go:
```diff
@@ -188,6 +195,7 @@ func (p *restrictedProfile) Apply(pod *corev1.Pod, containerName string, target
 
 	switch style {
 	case podCopy:
+		removeLabelsAndProbes(pod)
```
I agree with that but I think we need to manage this with on/off flag as we did in https://github.com/openshift/oc/blob/cf7b09cd29f1c0f56b2c6cd8193092c57c7730ff/pkg/cli/debug/debug.go#L252-L255 and default would be to removal of probes. WDYT? @verb

mochizuki875 replied:
@ardaguclu 

Thank you for your comment and I agree that.
`oc debug` seems to be implemented to be able to control the removal of probes, labels, annotations and initContainers.
On the other hands in `kubectl debug`, labels will be removed in all profiles and probes will be removed in some profiles(general and baseline).

So I'm wondering which should be targeted for removal control（only probes, probes and labels or all?）.
How do you think?

Current implementations of `kubectl debug` are as below.

#### probes
`removeLabelsAndProbes()` which remove probes already has been implemented and used in some profiles(general and baseline).
https://github.com/kubernetes/kubernetes/blob/91ee30074bee617d52fc24dc85132fe948aa5153/staging/src/k8s.io/kubectl/pkg/cmd/debug/profiles.go#L245-L254

#### labels
`removeLabelsAndProbes()` also implemented to remove labels, but copy pod labels will be removed when copy pod is created before  `removeLabelsAndProbes()` will be called.
https://github.com/kubernetes/kubernetes/blob/91ee30074bee617d52fc24dc85132fe948aa5153/staging/src/k8s.io/kubectl/pkg/cmd/debug/debug.go#L580-L589

#### annotations and initContainers
There seems to be not implemented the logic to remove annotations and initContainers.

mochizuki875 replied:
For now, I've added `--keep-*` flags to be able to control the removal of probes, labels, annotations and initContainers like `oc debug`.

mochizuki875 commented:
/remove-kind bug
/kind feature

mochizuki875 commented:
/retest

mochizuki875 commented:
PTAL?

ardaguclu commented on staging/src/k8s.io/kubectl/pkg/cmd/debug/debug.go:
```diff
@@ -708,6 +722,7 @@ func (o *DebugOptions) generatePodCopyWithDebugContainer(pod *corev1.Pod) (*core
 			Name:        o.CopyTo,
 			Namespace:   pod.Namespace,
 			Annotations: pod.Annotations,
+			Labels:      pod.Labels,
```
This is something unrelated to this change?

mochizuki875 replied:
Unless this change, original labels will be removed from copied and we can’t control keep or not them using --keep-labels flag.

ardaguclu commented on staging/src/k8s.io/kubectl/pkg/cmd/debug/debug_test.go:
```diff
@@ -346,7 +346,7 @@ func TestGenerateDebugContainer(t *testing.T) {
 			}
 
-			applier, err := NewProfileApplier(tc.opts.Profile)
+			applier, err := NewProfileApplier(tc.opts.Profile, &keepFields{})
```
I think, it would be better to use builder pattern. i.e.

```go
applier, err := NewProfileApplier(tc.opts.Profile)
if o.KeepAnnotations{
applier.WithKeepAnnotations()
}
...
```

mochizuki875 replied:
Thank you for your comment.
I've tried to fix.

ardaguclu commented on staging/src/k8s.io/kubectl/pkg/cmd/debug/profiles.go:
```diff
@@ -64,55 +64,56 @@ type ProfileApplier interface {
 	switch profile {
 	case ProfileLegacy:
-		return &legacyProfile{}, nil
+		return &legacyProfile{keeps}, nil
```
With the builder pattern, we don't need this

ardaguclu commented on staging/src/k8s.io/kubectl/pkg/cmd/debug/debug.go:
```diff
@@ -189,6 +196,12 @@ func (o *DebugOptions) AddFlags(cmd *cobra.Command) {
 	cmd.Flags().BoolVar(&o.Replace, "replace", o.Replace, i18n.T("When used with '--copy-to', delete the original Pod."))
 	cmd.Flags().StringToString("env", nil, i18n.T("Environment variables to set in the container."))
 	cmd.Flags().StringVar(&o.Image, "image", o.Image, i18n.T("Container image to use for debug container."))
+	cmd.Flags().BoolVar(&o.KeepLabels, "keep-labels", o.KeepLabels, "If true, keep the original pod labels")
```
Although there are 6 flags will be introduced (we don't usually prefer), I don't think there is any other alternative way to manage this. Besides, I think, these flags are essential for `kubectl debug` command.

ardaguclu reviewed (commented):
@mochizuki875 sorry for reviewing this after a long delay. These changes look good to me and I agree with you, these are essential. Just left some comments.

ardaguclu commented:
/triage accepted
/priority backlog

ardaguclu commented on staging/src/k8s.io/kubectl/pkg/cmd/debug/profiles.go:
```diff
@@ -210,6 +245,10 @@ func (p *netadminProfile) Apply(pod *corev1.Pod, containerName string, target ru
 		useHostNamespaces(pod)
 
 	case podCopy:
+		p.keeps.removeLabels(pod)
```
and this flags are only applicable for copying pod debug type and not relevant for ephemeral debug and node debug. I'm thinking maybe we should mention in the flags descriptions to this.

mochizuki875 replied:
I've added the flags descriptions like "This flag only works when used with '--copy-to'".

ardaguclu commented:
This PR fixes https://github.com/kubernetes/kubectl/issues/1570

mochizuki875 commented:
@ardaguclu 
Thank you for your RVs and I've fixed them.
Please check again.

ardaguclu commented on staging/src/k8s.io/kubectl/pkg/cmd/debug/profiles.go:
```diff
@@ -61,6 +61,13 @@ const (
 type ProfileApplier interface {
 	// Apply applies the profile to the given container in the pod.
 	Apply(pod *corev1.Pod, containerName string, target runtime.Object) error
+	NewKeepLabels()
```
Instead of extending the interface just for this, I was thinking something similar;

```diff
diff --git a/staging/src/k8s.io/kubectl/pkg/cmd/debug/profiles.go b/staging/src/k8s.io/kubectl/pkg/cmd/debug/profiles.go
index 2e0f8621ee7..863fc07bd78 100644
--- a/staging/src/k8s.io/kubectl/pkg/cmd/debug/profiles.go
+++ b/staging/src/k8s.io/kubectl/pkg/cmd/debug/profiles.go
@@ -58,13 +58,64 @@ const (
        ProfileSysadmin = "sysadmin"
 )
 
+type KeepFields struct {
+       labels         bool
+       annotations    bool
+       liveness       bool
+       readiness      bool
+       startup        bool
+       initContainers bool
+}
+
+func NewKeepFields() KeepFields {
+       return KeepFields{
+               initContainers: true,
+       }
+}
+
+// removeLabels removes labels from the pod.
+func (keeps KeepFields) removeLabels(p *corev1.Pod) {
+       if !keeps.labels {
+               p.Labels = nil
+       }
+}
+
+// removeAnnotations remove annotations from the pod.
+func (keeps KeepFields) removeAnnotations(p *corev1.Pod) {
+       if !keeps.annotations {
+               p.Annotations = nil
+       }
+}
+
+// removeProbes remove probes from all containers of the pod.
+func (keeps KeepFields) removeProbes(p *corev1.Pod) {
+       for i := range p.Spec.Containers {
+               if !keeps.liveness {
+                       p.Spec.Containers[i].LivenessProbe = nil
+               }
+               if !keeps.readiness {
+                       p.Spec.Containers[i].ReadinessProbe = nil
+               }
+               if !keeps.startup {
+                       p.Spec.Containers[i].StartupProbe = nil
+               }
+       }
+}
+
+// removeInitContainers remove initContainers from the pod.
+func (keeps KeepFields) removeInitContainers(p *corev1.Pod) {
+       if !keeps.initContainers {
+               p.Spec.InitContainers = nil
+       }
+}
+
 type ProfileApplier interface {
        // Apply applies the profile to the given container in the pod.
        Apply(pod *corev1.Pod, containerName string, target runtime.Object) error
 }
 
 // NewProfileApplier returns a new Options for the given profile name.
-func NewProfileApplier(profile string) (ProfileApplier, error) {
+func NewProfileApplier(profile string, keepFields KeepFields) (ProfileApplier, error) {
        switch profile {
        case ProfileLegacy:
                return &legacyProfile{}, nil
@@ -87,12 +138,15 @@ type legacyProfile struct {
 }
 
 type generalProfile struct {
+       KeepFields
 }
 
 type baselineProfile struct {
+       KeepFields
 }
 
 type restrictedProfile struct {
+       KeepFields
 }
 
 type netadminProfile struct {
@@ -164,7 +218,7 @@ func (p *baselineProfile) Apply(pod *corev1.Pod, containerName string, target ru
 
        switch style {
        case podCopy:
-               removeLabelsAndProbes(pod)
+               p.KeepFields.removeAnnotations(pod)
                shareProcessNamespace(pod)
 
        case ephemeral, node:
@@ -242,17 +296,6 @@ func (p *sysadminProfile) Apply(pod *corev1.Pod, containerName string, target ru
        return nil
 }
 
-// removeLabelsAndProbes removes labels from the pod and remove probes
-// from all containers of the pod.
-func removeLabelsAndProbes(p *corev1.Pod) {
-       p.Labels = nil
-       for i := range p.Spec.Containers {
-               p.Spec.Containers[i].LivenessProbe = nil
-               p.Spec.Containers[i].ReadinessProbe = nil
-               p.Spec.Containers[i].StartupProbe = nil
-       }
-}
-

```

ardaguclu commented on staging/src/k8s.io/kubectl/pkg/cmd/debug/debug.go:
```diff
@@ -261,6 +274,25 @@ func (o *DebugOptions) Complete(restClientGetter genericclioptions.RESTClientGet
 		if err != nil {
 			return err
 		}
+		applier.NewKeepLabels()
```
instead of all these functions;

```go
applier, err := NewProfileApplier(o.Profile, NewKeepFields())
```
and of course the flags will replace the values in `NewKeepFields`?

mochizuki875 replied:
I understood you intend not to change NewProfileApplier()  interface from [previous comment](https://github.com/kubernetes/kubernetes/pull/123149#discussion_r1591880343).
Is it ok to change it like `NewProfileApplier(ProfileApplier, KeepFields)`?

ardaguclu replied:
yes, it is ok because it has no impact on the interface signature. It just changes the `NewProfileApplier` function's signature which should be less impactful.

ardaguclu replied:
Yeah, your previous version was less disruptive than the current one.

ardaguclu commented on staging/src/k8s.io/kubectl/pkg/cmd/debug/profiles.go:
```diff
@@ -130,10 +157,34 @@ func getDebugStyle(pod *corev1.Pod, target runtime.Object) (debugStyle, error) {
+		useHostNamespaces(pod)
+
+	case podCopy:
+		p.keeps.removeLabels(pod)
```
I think this is a behavioral change for legacyProfile because we were not removing as default and now we are.
I think, we can leave it as is and even keep flags don't change it. As it is a legacy profile.

mochizuki875 replied:
Do you mean copy pod applied legacyProfile should keep annotations, probes and initContainers and doesn’t  keep labels(it is not kept now as mentioned [here](https://github.com/kubernetes/kubernetes/pull/123149#discussion_r1596918676)) regardless of --keep-* flags?

ardaguclu replied:
Yes that is true. Because it is a legacy profile anyway and we shouldn't change the current behavior, unless it is fully broken, etc.

mochizuki875 replied:
Thanks!
So should we add additional flags descriptions like "it does not work in legacyProfile" related [that](https://github.com/kubernetes/kubernetes/pull/123149#discussion_r1596976171)?

ardaguclu replied:
Let's don't add it for now. I think, adding more text just for a a legacy profile seems to be unnecessary. Isn't the recommended path is using the other profiles other than `legacy`?

ardaguclu replied:
We can say that if you want to manage these resources, please use the other recommended paths.

mochizuki875 replied:
> Let's don't add it for now. 

Ok, I've got it.

> Isn't the recommended path is using the other profiles other than legacy?

By the way, how many users recognize it recommended to use other profiles than legacy profile and the existence of debugging profile?

- The official document doesn't seem to mention debugging profiles.(It's mentioned only in [KEP](https://github.com/kubernetes/enhancements/blob/master/keps/sig-cli/1441-kubectl-debug/README.md#debugging-profiles) now as I know.)
- "kubectl debug --help" doesn't seem to show any recommended paths regarding profiles and show default profile is legacy.

So IMO, I think many users don't know the existence of debugging profile and recommendation.
Shouldn't we add any description about debugging profile somewhere?

In addition, it is said that in [KEP](https://github.com/kubernetes/enhancements/blob/master/keps/sig-cli/1441-kubectl-debug/README.md#default-profile-and-automation-selection):

> Default Profile and Automation Selection
In order to provide a seamless experience and encourage use of [PodSecurity](http://kep.k8s.io/2579), the "auto" profile will automatically choose a profile that's compatible with the current security profile by examining the pod-security.kubernetes.io/enforce annotation on the namespace and selecting the most permissive of "general", "baseline", and "restricted" that the controller will allow.
> 
> This will become the default behavior, but in order to maintain backwards compatibility the "legacy" profile will be the default profile until the 1.25 release. When --profile is not specified kubectl debug will print a warning about the upcoming change in behavior.

Though auto profile is WIP(#120687) now, shouldn't we print any warning or recommendation about profile as a first step?

These are outside the scope of this PR, but I was a little curious.

ardaguclu replied:
I assume that `legacy` name is self explanatory but I see your comments and it all makes sense. We should reconsider what is the path for legacy profile.

ardaguclu commented on staging/src/k8s.io/kubectl/pkg/cmd/debug/profiles.go:
```diff
@@ -188,6 +245,10 @@ func (p *restrictedProfile) Apply(pod *corev1.Pod, containerName string, target
 
 	switch style {
 	case podCopy:
+		p.keeps.removeLabels(pod)
```
As far as I understand, this is a missing bit for restricted profile and netadmin profile and sysadmin profile. Because I don't know any reason of why we don't clear labels and probes in these profiles. And it seems that we are also fixing this.

ardaguclu reviewed (commented):
Thanks for spending time and effort on this PR. That would be great, if we can add some integration tests cases also in here https://github.com/kubernetes/kubernetes/blob/master/test/cmd/debug.sh

mochizuki875 commented:
@ardaguclu 
Ok, I've fixed.

> Thanks for spending time and effort on this PR. That would be great, if we can add some integration tests cases also in here https://github.com/kubernetes/kubernetes/blob/master/test/cmd/debug.sh

I've added test pod manifest(`hack/testdata/pod-with-metadata-and-probes.yaml`) for added test cases.
If there is some good idea instead of adding the manifest, please comment that.

mochizuki875 commented:
/retest

ardaguclu commented on staging/src/k8s.io/kubectl/pkg/cmd/debug/profiles.go:
```diff
@@ -64,54 +64,121 @@ type ProfileApplier interface {
+	initContainers bool
+}
+
+func NewKeepFields() *KeepFields {
```
nit: some godoc for exported types

ardaguclu commented on staging/src/k8s.io/kubectl/pkg/cmd/debug/profiles.go:
```diff
@@ -64,54 +64,121 @@ type ProfileApplier interface {
-		return nil
-	default:
-		return fmt.Errorf("the %s profile doesn't support objects of type %T", ProfileLegacy, target)
+type KeepFields struct {
```
nit: godoc

ardaguclu commented on staging/src/k8s.io/kubectl/pkg/cmd/debug/profiles.go:
```diff
@@ -64,54 +64,121 @@ type ProfileApplier interface {
+}
+
+// removeLabels removes labels from the pod.
+func (keeps *KeepFields) removeLabels(p *corev1.Pod) {
```
`KeepFields` is an exported type but it's all actual functions are not exported seems to be counter-intuitive. I know they are not used by anywhere else but what do you think about exporting the functions as well?

mochizuki875 replied:
The fix would likely eliminate the need to export types and functions in the first place.

ardaguclu commented on staging/src/k8s.io/kubectl/pkg/cmd/debug/debug.go:
```diff
@@ -257,7 +270,8 @@ func (o *DebugOptions) Complete(restClientGetter genericclioptions.RESTClientGet
 
 	if o.Applier == nil {
-		applier, err := NewProfileApplier(o.Profile)
+		keeps := NewKeepFields().WithKeepLabels(o.KeepLabels).WithKeepAnnotations(o.KeepAnnotations).WithKeepLiveness(o.KeepLiveness).WithKeepReadiness(o.KeepReadiness).WithKeepStartup(o.KeepStartup).WithKeepInitContainers(o.KeepInitContainers)
```
Instead of all these functions, what about;
```suggestion
		keeps := &KeepFields{
		Labels: o.KeepLabels,
		Annotations: o.KeepAnnotations
		...
		}
```
I know I was the suggester of builder pattern but unfortunately it seems redundant now as `ProfileApplier` is actually an interface and we don't want to change it's signature.

ardaguclu commented on test/cmd/debug.sh:
```diff
@@ -51,6 +49,26 @@ run_kubectl_debug_pod_tests() {
+  kube::test::get_object_assert pod/target-copy '{{.metadata.annotations}}' 'map\[test:test\]'
+  kube::test::get_object_assert pod/target-copy '{{range.spec.containers}}{{.name}}:{{end}}' 'target:debug-container:'
+  kube::test::get_object_assert pod/target-copy '{{range.spec.containers}}{{.image}}:{{end}}' "${IMAGE_NGINX:?}:busybox:"
+  kube::test::get_object_assert pod/target-copy '{{range.spec.containers}}{{if (index . "livenessProbe")}}:{{end}}{{end}}' ':'
```
What is the expected output in here and why are we checking `:`?

mochizuki875 replied:
It checks whether any containers in target-copy has livenessProbe.
If it dose, go-template return `:`.
I reused an existing template:
https://github.com/kubernetes/kubernetes/blob/8362f51c6620a55953ac564d24fa82d2f2ea2403/test/cmd/debug.sh#L147C1-L147C126

However, if it is not intuitive, I'll fix it.

eg1: return `true` if container has livenessProbe.
```sh
kube::test::get_object_assert pod/target-copy '{{range.spec.containers}}{{if (index . "livenessProbe")}}true{{end}}{{end}}' 'true'
```

eg2: return livenessProbe's exec field value if container has livenessProbe and check them.
```sh
kube::test::get_object_assert pod/target-copy '{{range.spec.containers}}{{.livenessProbe.exec}}:{{end}}' 'map\[command:\[/bin/sh -c cat probe\]\]:<no value>:'
```

ardaguclu reviewed (commented):
tests look great. Thanks.

mochizuki875 commented:
@ardaguclu
I've fixed.
In addition, I've renamed `KeepFields` to `keepFlags` because previous name seems to be not intuitive.

ardaguclu commented on staging/src/k8s.io/kubectl/pkg/cmd/debug/debug.go:
```diff
@@ -257,7 +270,15 @@ func (o *DebugOptions) Complete(restClientGetter genericclioptions.RESTClientGet
 
 	if o.Applier == nil {
-		applier, err := NewProfileApplier(o.Profile)
+		kflags := keepFlags{
```
Agree that `keepFlags` is better naming and sorry, if I wasn't clear enough: Could you please export this, it's functions and it's fields?.

mochizuki875 replied:
> Could you please export this, it's functions and it's fields?.

It's ok.
Could you tell me the reason?
As you said [here](https://github.com/kubernetes/kubernetes/pull/123149#discussion_r1602072815), they are not used by anywhere so I thought we need not to export...

ardaguclu replied:
> > Could you please export this, it's functions and it's fields?.
> 
> It's ok. Could you tell me the reason? As you said [here](https://github.com/kubernetes/kubernetes/pull/123149#discussion_r1602072815), they are not used by anywhere so I thought we need not to export...

Basically, in case some other tools depend on kubectl debug functionality, they would be able to customize the behavior. Otherwise, keep flag functionality will be black box. We can also may want to intentionally unexport them. However, they would fully be relying on the flags and it may better to export them in my opinion. Hope that it makes sense.

ardaguclu replied:
Another reason would be `ProfileApplier` is exported function and can be customizable. On the other hand, it only gets an unexported type which drastically hurts the usability of `ProfileApplier`.

mochizuki875 replied:
Thank you for your explanation.
I've understood.

ardaguclu reviewed (commented):
One last comment and after that I think we are good to go

ardaguclu commented on staging/src/k8s.io/kubectl/pkg/cmd/debug/profiles.go:
```diff
@@ -210,6 +279,11 @@ func (p *netadminProfile) Apply(pod *corev1.Pod, containerName string, target ru
+		p.removeAnnotations(pod)
+		p.removeProbes(pod)
+		p.removeInitContainers(pod)
+		p.removeInitContainers(pod)
```
This is duplicate

mochizuki875 replied:
Thank you and I've noticed just now.
I've removed it.

mochizuki875 commented:
@ardaguclu 
I've fixed and rebased.

mochizuki875 commented on staging/src/k8s.io/kubectl/pkg/cmd/debug/profiles.go:
```diff
@@ -64,54 +64,91 @@ type ProfileApplier interface {
-		return fmt.Errorf("the %s profile doesn't support objects of type %T", ProfileLegacy, target)
+// KeepFlags holds the flag set that determine which fields to keep in the copy pod.
+type KeepFlags struct {
+	labels         bool
```
Sorry, I overlooked it.
I've fixed.

ardaguclu commented:
Great, thanks!
/approve
/lgtm

mochizuki875 commented:
@ardaguclu 
I'm very appreciate your many comments!
